From 734eb0d461b77e8c33aa95590e008c8c287990c2 Mon Sep 17 00:00:00 2001
From: Alexander Mikhalitsyn <aleksandr.mikhalitsyn@canonical.com>
Date: Sun, 6 Apr 2025 20:42:26 +0200
Subject: [PATCH] criu/mem: dump: skip MADV_GUARD pages content dump

1. get info about MADV_GUARD_INSTALL-protected pages with
help of pagemap by looking for PME_GUARD_REGION flag if /proc/<pid>/pagemap
is used or by looking for PAGE_IS_GUARD flag if ioctl(PAGEMAP_SCAN) is used
2. skip those pages

Signed-off-by: Alexander Mikhalitsyn <aleksandr.mikhalitsyn@canonical.com>
(cherry picked from commit cc047d595f742e416220d2d7740334500eb96a85)

Conflicts:
	criu/mem.c
---
 criu/include/mem.h | 1 +
 criu/mem.c         | 9 +++++++++
 2 files changed, 10 insertions(+)

diff --git a/criu/include/mem.h b/criu/include/mem.h
index 3618c9cc3..184acd024 100644
--- a/criu/include/mem.h
+++ b/criu/include/mem.h
@@ -35,6 +35,7 @@ extern int parasite_dump_pages_seized(struct pstree_item *item, struct vm_area_l
 #define PME_PRESENT	  (1ULL << 63)
 #define PME_SWAP	  (1ULL << 62)
 #define PME_FILE	  (1ULL << 61)
+#define PME_GUARD_REGION  (1ULL << 58)
 #define PME_SOFT_DIRTY	  (1ULL << 55)
 #define PME_PSHIFT_BITS	  (6)
 #define PME_STATUS_BITS	  (3)
diff --git a/criu/mem.c b/criu/mem.c
index f4dd684ba..562abf961 100644
--- a/criu/mem.c
+++ b/criu/mem.c
@@ -135,12 +135,19 @@ u64 should_dump_page(pmc_t *pmc, VmaEntry *vmae, u64 vaddr, bool *softdirty)
 		}
 		if (vaddr < pmc->regs[pmc->regs_idx].start)
 			return pmc->regs[pmc->regs_idx].start;
+
+		if (pmc->regs[pmc->regs_idx].categories & PAGE_IS_GUARD)
+			goto skip_guard_page;
+
 		if (softdirty)
 			*softdirty = pmc->regs[pmc->regs_idx].categories & PAGE_IS_SOFT_DIRTY;
 		return vaddr;
 	} else {
 		u64 pme = pmc->map[PAGE_PFN(vaddr - pmc->start)];
 
+		if (pme & PME_GUARD_REGION)
+			goto skip_guard_page;
+
 		/*
 		 * Optimisation for private mapping pages, that haven't
 		 * yet being COW-ed
@@ -155,6 +162,8 @@ u64 should_dump_page(pmc_t *pmc, VmaEntry *vmae, u64 vaddr, bool *softdirty)
 
 		return vaddr + PAGE_SIZE;
 	}
+skip_guard_page:
+	return vaddr + PAGE_SIZE;	
 }
 
 bool page_is_zero(u64 pme)
-- 
2.34.1

